netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -109,160 +110,180 @@ static NettyServerHandler newHandler(ServerTransportListener transportListener,
+
+    return newHandler(builder, transportListener, maxStreams, flowControlWindow, maxHeaderListSize,
         maxMessageSize);
   }
 
   @VisibleForTesting
-  static NettyServerHandler newHandler(Http2FrameReader frameReader, Http2FrameWriter frameWriter,
+  static NettyServerHandler newHandler(Http2FrameCodecBuilder frameCodecBuilder,
                                        ServerTransportListener transportListener,
                                        int maxStreams,
                                        int flowControlWindow,
+                                       int maxHeaderListSize,
                                        int maxMessageSize) {
     Preconditions.checkArgument(maxStreams > 0, "maxStreams must be positive");
     Preconditions.checkArgument(flowControlWindow > 0, "flowControlWindow must be positive");
     Preconditions.checkArgument(maxMessageSize > 0, "maxMessageSize must be positive");
 
-    Http2Connection connection = new DefaultHttp2Connection(true);
-
-    // Create the local flow controller configured to auto-refill the connection window.
-    connection.local().flowController(
-        new DefaultHttp2LocalFlowController(connection, DEFAULT_WINDOW_UPDATE_RATIO, true));
-
+    frameCodecBuilder.initialSettings(
+        new Http2Settings().initialWindowSize(flowControlWindow).maxConcurrentStreams(maxStreams)
+            .maxHeaderListSize(maxHeaderListSize));
 
-    Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-    Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder,
-        frameReader);
-
-    Http2Settings settings = new Http2Settings();
-    settings.initialWindowSize(flowControlWindow);
-    settings.maxConcurrentStreams(maxStreams);
-
-    return new NettyServerHandler(transportListener, decoder, encoder, settings, maxMessageSize);
+    return new NettyServerHandler(frameCodecBuilder, flowControlWindow,
+        transportListener, maxMessageSize);
   }
 
-  private NettyServerHandler(ServerTransportListener transportListener,
-                             Http2ConnectionDecoder decoder,
-                             Http2ConnectionEncoder encoder, Http2Settings settings,
-                             int maxMessageSize) {
-    super(decoder, encoder, settings);
+  private NettyServerHandler(Http2FrameCodecBuilder frameCodecBuilder, int flowControlWindow,
+      ServerTransportListener transportListener, int maxMessageSize) {
+    super(frameCodecBuilder, flowControlWindow);
     checkArgument(maxMessageSize >= 0, "maxMessageSize must be >= 0");
     this.maxMessageSize = maxMessageSize;
-
-    streamKey = encoder.connection().newKey();
     this.transportListener = checkNotNull(transportListener, "transportListener");
+  }
 
-    // Set the frame listener on the decoder.
-    decoder().frameListener(new FrameListener());
+  @Override
+  public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
```

Comment:
This need to propagate the messages or make sure to release them (reference counting). Right now any unknown message is leaked.